File: 1035.Uncrossed Lines.py
class Solution(object):
    def maxUncrossedLines(self, A, B):
        """
        :type A: List[int]
        :type B: List[int]
        :rtype: int
        """
        a, b = len(A), len(B)
        dp = [[0] * b for _ in xrange(a)]
        dp[0][0] = int(A[0] == B[0])
        for i in xrange(1, a):
            dp[i][0] = dp[i - 1][0] or int(A[i] == B[0])
        for j in xrange(1, b):
            dp[0][j] = dp[0][j - 1] or int(A[0] == B[j])
        for i in xrange(1, a):
            for j in xrange(1, b):
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
                if A[i] == B[j]:
                    dp[i][j] = max(dp[i][j], dp[i - 1][j - 1] + 1)
        return dp[-1][-1]

File: 1001-1500/1035.Uncrossed Lines.rb
# @param {Integer[]} a
# @param {Integer[]} b
# @return {Integer}
def max_uncrossed_lines(a, b)
    @dp, @a, @b = Hash.new(0), a, b
    def search(i, j)
        return 0 if i < 0 || j < 0
        if !@dp.include? [i, j]
            @dp[[i, j]] = search(i - 1, j - 1) + 1 if @a[i] == @b[j]
            @dp[[i, j]] = [@dp[[i, j]], search(i - 1, j), search(i, j - 1)].max
        end
        @dp[[i, j]]
    end
    search(a.size - 1, b.size - 1)
end
